webknossos-datastore/app/com/scalableminds/webknossos/datastore/explore/NgffV0_5Explorer.scala

```diff
@@ -0,0 +1,156 @@
+                                       channelCount: Int,
+                                       channelAttributes: Option[Seq[ChannelAttributes]] = None,
+                                       isSegmentation: Boolean = false): Fox[List[(Zarr3Layer, VoxelSize)]] =
+    for {
+      axisOrder <- extractAxisOrder(multiscale.axes) ?~> "Could not extract XYZ axis order mapping. Does the data have x, y and z axes, stated in multiscales metadata?"
+      unifiedAxisUnit <- selectAxisUnit(multiscale.axes, axisOrder)
+      axisUnitFactors <- extractAxisUnitFactors(unifiedAxisUnit, multiscale.axes, axisOrder) ?~> "Could not extract axis unit-to-nm factors"
+      voxelSizeInAxisUnits <- extractVoxelSizeInAxisUnits(
+        multiscale.datasets.map(_.coordinateTransformations),
+        axisOrder) ?~> "Could not extract voxel size from scale transforms"
+      voxelSizeFactor = voxelSizeInAxisUnits * axisUnitFactors
+      nameFromPath = guessNameFromPath(remotePath)
+      name = multiscale.name.getOrElse(nameFromPath)
+      layerTuples <- Fox.serialCombined((0 until channelCount).toList)({ channelIndex: Int =>
+        for {
+          magsWithAttributes <- Fox.serialCombined(multiscale.datasets)(d =>
+            zarrMagFromNgffDataset(d, remotePath, voxelSizeInAxisUnits, axisOrder, credentialId, Some(channelIndex)))
+          _ <- bool2Fox(magsWithAttributes.nonEmpty) ?~> "zero mags in layer"
+          elementClassRaw <- elementClassFromMags(magsWithAttributes) ?~> "Could not extract element class from mags"
+          elementClass = if (isSegmentation) ensureElementClassForSegmentationLayer(elementClassRaw)
+          else elementClassRaw
+
+          (viewConfig: LayerViewConfiguration, channelName: String) = parseChannelAttributes(channelAttributes,
+                                                                                             name,
+                                                                                             channelIndex)
+
+          boundingBox = boundingBoxFromMags(magsWithAttributes)
+          additionalAxes <- getAdditionalAxes(multiscale, remotePath)
+          layer: Zarr3Layer = if (looksLikeSegmentationLayer(name, elementClass) || isSegmentation) {
+            Zarr3SegmentationLayer(
+              channelName,
+              boundingBox,
+              elementClass,
+              magsWithAttributes.map(_.mag),
+              largestSegmentId = None,
+              additionalAxes = Some(additionalAxes),
+              defaultViewConfiguration = Some(viewConfig),
+            )
+          } else
+            Zarr3DataLayer(
+              channelName,
+              Category.color,
+              boundingBox,
+              elementClass,
+              magsWithAttributes.map(_.mag),
+              additionalAxes = Some(additionalAxes),
+              defaultViewConfiguration = Some(viewConfig),
+            )
+        } yield (layer, VoxelSize(voxelSizeFactor, unifiedAxisUnit))
```

Comment:
looks like there is still quite a bit of duplication between 0_5 and 0_4. I wonder if some more code could be extracted into functions to be moved into the new trait? But I might have overlooked some subtle differences.

Reply:
zarrMagFromNgffDataset needs to remain duplicated because it uses getZarrHeader, which returns differnt headers (case classes, so I don't think we can use a trait here).

But I moved a liitle more.